pmlc/dialect/abi/transforms/lower_to_abi.cc

```diff
@@ -0,0 +1,132 @@
+  auto loopOp = builder.create<abi::LoopOp>(loc);
+
+  // Initialize the loop's body by taking the main function's body.
+  loopOp.bodyRegion().takeBody(mainFunc.getBody());
+
+  // A function's blocks can end with std.return; since std.return expects its
+  // parent to be std.func, we need to replace any cloned top-level std.return
+  // operations with abi.done.
+  for (auto &op : llvm::make_early_inc_range(loopOp.bodyRegion().getOps())) {
+    auto returnOp = mlir::dyn_cast<mlir::ReturnOp>(op);
+    if (returnOp) {
+      builder.setInsertionPoint(returnOp);
+      builder.create<abi::TerminatorOp>(returnOp.getLoc());
+      returnOp.erase();
+    }
+  }
+
+  // Add an entry block for the init region.
+  auto *initEntryBlock = builder.createBlock(&loopOp.initRegion());
+
+  // Non-constant arguments will vary from invocation to invocation; they are
+  // intrinsically parameters of the body entry block.  Constant arguments do
+  // not vary; we pass them to the initializtion block, which can then pass them
+  // on to the body as needed.
+  //
+  // At this point, all of the arguments should be memrefs, except for a
+  // possible initial device argument -- which, if present, is always passed
+  // to the initialization block.
+  auto *bodyEntryBlock = loopOp.getBodyEntryBlock();
+  unsigned argIdx = 0;
+  mlir::Identifier constAttrId = builder.getIdentifier("tile.const");
+  if (bodyEntryBlock->getNumArguments() == 0 ||
+      bodyEntryBlock->getArgument(0)
+          .getType()
+          .isa<mlir::MemRefType, mlir::UnrankedMemRefType>()) {
+    // The body entry block started with a memref (or didn't have any
+    // arguments --we should always have at least *one* argument, but we handle
+    // the no-argument case for completeness).  So, create a fake device
+    // parameter.
+    initEntryBlock->addArgument(
+        LLVMType::getInt8Ty(&getContext()).getPointerTo());
+  }
+
+  mlir::SmallVector<mlir::Value, 8> networkArgs;
+  mlir::SmallVector<mlir::Type, 8> networkTypes;
+  while (argIdx < bodyEntryBlock->getNumArguments()) {
+    auto arg = bodyEntryBlock->getArgument(argIdx);
+    bool isMemRef =
+        arg.getType().isa<mlir::MemRefType, mlir::UnrankedMemRefType>();
+    if (argIdx && !isMemRef) {
+      mainFunc.emitError(
+          "Expected only memrefs after an optional device parameter");
+      signalPassFailure();
+      return;
+    }
+    auto constAttr =
+        mainFunc.getArgAttrOfType<mlir::IntegerAttr>(argIdx, constAttrId);
+    if (constAttr || (!argIdx && !isMemRef)) {
+      auto ty = arg.getType();
+      auto newArg = bodyEntryBlock->insertArgument(networkArgs.size(), ty);
+      arg.replaceAllUsesWith(newArg);
+      bodyEntryBlock->eraseArgument(argIdx + 1);
+      initEntryBlock->addArgument(ty);
+      networkArgs.emplace_back(
+          initEntryBlock->getArgument(initEntryBlock->getNumArguments() - 1));
+      networkTypes.emplace_back(ty);
+    }
+    ++argIdx;
```

Comment:
Use `llvm::enumerate` instead of tracking the argIdx independently?

Reply:
Actually, you can just use `bodyEntryBlock->getArgNumber()`

Reply:
Done; thanks!